pilot/pkg/serviceregistry/external/servicediscovery.go

```diff
@@ -52,36 +51,67 @@ type ServiceEntryStore struct {
 }
 
 // NewServiceDiscovery creates a new ServiceEntry discovery service
-func NewServiceDiscovery(configController model.ConfigStoreCache, store model.IstioConfigStore) *ServiceEntryStore {
+func NewServiceDiscovery(configController model.ConfigStoreCache, store model.IstioConfigStore, xdsUpdater model.XDSUpdater) *ServiceEntryStore {
 	c := &ServiceEntryStore{
-		serviceHandlers:  make([]serviceHandler, 0),
-		instanceHandlers: make([]instanceHandler, 0),
-		store:            store,
-		ip2instance:      map[string][]*model.ServiceInstance{},
-		instances:        map[host.Name]map[string][]*model.ServiceInstance{},
-		updateNeeded:     true,
+		XdsUpdater:   xdsUpdater,
+		store:        store,
+		ip2instance:  map[string][]*model.ServiceInstance{},
+		instances:    map[host.Name]map[string][]*model.ServiceInstance{},
+		updateNeeded: true,
 	}
 	if configController != nil {
-		configController.RegisterEventHandler(schemas.ServiceEntry.Type, func(config model.Config, event model.Event) {
+		configController.RegisterEventHandler(schemas.ServiceEntry.Type, func(old, curr model.Config, event model.Event) {
 			// Recomputing the index here is too expensive.
 			c.changeMutex.Lock()
 			c.lastChange = time.Now()
 			c.updateNeeded = true
 			c.changeMutex.Unlock()
 
-			// TODO : Currently any update to ServiceEntry triggers a full push. We need to identify what has actually
-			// changed and call appropriate handlers - for example call only service handler for the changed services
-			// and call instance handlers only when instance update happens. This requires us to rework the handlers to
-			// have both old and new objects so that they can compare and be smart.
-			services := convertServices(config)
-			for _, handler := range c.serviceHandlers {
-				for _, service := range services {
-					go handler(service, event)
+			cs := convertServices(curr)
```

Comment:
This is the main change, Rest of the change is plumbing the old + curr objects on all the event handlers